### PR TITLE
Add Python DB Logger

### LINKED ISSUE #134: Incorporate `python_db_logger`
Add `python_db_logger` to Nacculator:
- Add an `example.env` with the necessary `DB_*` environment variables
- Add a local mysql container definition. Copy the folder structure (`.containers/mysql`) and `docker-compose.yml` from [python_db_logger](https://github.com/ctsit/python_db_logger/blob/master/.containers/mysql/docker-compose.yml)
- Update README with instructions on how to start the docker container
- Add `python_db_logger` as a dependency in `setup.py`:
> ```python
> install_requires=[
>       "python_db_logger @ git+ssh://git@github.com:/ctsit/python_db_logger.git"
>   ]
>```
- Create a `configure_logging` function similar to https://github.com/ctsit/ufvivo-pubingest/blob/0779cf62b41245c479ebf49a9233d77bd644ab58/ufvivo/pubingest/__main__.py#L89-L111 that also creates an instance of `DBLogger`
- Replace all calls to `print` with `db_logger.print_info()` , `db_logger.print_warn()` , etc.

### PR DESCRIPTION
**Pull request by rtdtwo**
## What does this change do?
This change adds Python DB Logger module to Nacculator, to log directly to our logging database.

## Why was this change made?
This change fixes #134 

## Verification
1. First, copy the example env file:
`cp db_logger.env.example .env`
2. Then, start a MySQL instance. This will create the database and the relevant tables and set up Adminer for easy database visualization.
`docker compose -f .containers/mysql/docker-compose.yml up -d`
3. Run Nacculator with the flags and configurations of your choice.
4. Once Nacculator has run successfully, you should find a `report.xls` file in Nacculator's root directory.
5. Open the XLS file and verify that the expected logs (Success, Skip, Failure) are logged.
6. Next, go to Adminer (http://localhost:8080). DB logger database is already configured. Login with the database credentials found in the env file.
7. Verify the table `detail_log` has the logs.

Both the database and XLS will have the same logs you would expect from the stdout and stderr file outputs.

No other configurations are needed, DB Logger has already been set up to start logging automatically when Nacculator script starts running.

## Affirmations
* [x] I matched the style of the existing code.
* [x] I added and updated any relevant documentation (inline, `README`, `CHANGELOG`, and such).
* [x] I used Python's type hinting.
* [x] I ran the automated tests and ensured they **ALL** passed.
* [x] I ran the linter and ensured my changes have not introduced **ANY** warnings or errors.
* [x] I have made an effort to minimize code changes and have not included any cruft (preference files, *.pyc files, old comments, print-debugging, unused variables).
* [x] I have made an effort maintain a clear commit history (haven't merged other branches or rebased improperly).
* [x] I have written the code myself or have given credit where credit is due.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 
     install_requires=[
         "PyCap>=2.1.0"
```
Comment (s-emerson) [inline setup.py:35]: needs a comma i think

**Comment by s-emerson** [comment]
I am unable to install the db_logger package using the directions in https://github.com/ctsit/python_db_logger . I'll need help getting this set up and working when you get a chance, so that I can test the code.

File: setup.py (hunk 1 above)
Comment (rtdtwo) [reply setup.py]: This was one of the files which never got committed, so this shouldn't have been here at the first place. This fixes python_db_logger not getting installed when you do a `pip install .`

**Comment by rtdtwo** [comment]
@s-emerson Everything works as expected in a local run.

File: .containers/mysql/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+version: '3.1'
+
+services:
+
+  db:
```
Comment (s-emerson) [inline .containers/mysql/docker-compose.yml:5]: @mbentz-uf can you add phpmyadmin and default schema to this? i can't test the code by running it using the steps provided in the pull request.

File: db_logger.env.example
Comment (s-emerson) [inline db_logger.env.example:1]: is there supposed to be a second example env file?

File: .containers/mysql/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    image: mysql:8.0.29
+    restart: always
+    environment:
+      MYSQL_DATABASE: log_db
+      MYSQL_USER: dev
+      MYSQL_PASSWORD: password
+      MYSQL_ROOT_PASSWORD: password
+    ports:
+      - 4000:3306
```
[the remark this answers is not in the document]
Comment (rtdtwo) [reply .containers/mysql/docker-compose.yml:13]: Done!

**Comment by rtdtwo** [comment]
@s-emerson I have added adminer which should be accessible at http://localhost:8080 once the containers are up. Docker should also automatically create all tables. A local run confirms all of this works. Please try running Nacculator again and let me know if you face any issues.

File: db_logger.env.example
Comment (rtdtwo) [reply db_logger.env.example:1]: Keeping this and deleting the other one.

File: .containers/mysql/docker-compose.yml (hunk 2 above)
Comment (rtdtwo) [reply .containers/mysql/docker-compose.yml:4]: Done!

**Comment by rtdtwo** [comment]
@s-emerson Updated the details steps in "verification" section of the pull request.

**Comment by s-emerson** [comment]
i was able to run nacculator with dblogger installed, though i still had issues with the local-only installation that i think will make this fail for anyone who is not part of uf (especially with python_db_logger as a private repo). 

the logging is not keeping track of warnings that nacculator catches. I have attached screencaps of the regular error output vs the logged csv file. 
Screen Shot 2023-04-12 at 11 31 25 AM
Screen Shot 2023-04-12 at 11 31 15 AM

i am also still getting an issue in the output ivp.txt meant to submit to nacc-
"INSERT INTO detail_log (script_name, script_run_time, message, level, process_id)VALUES('db_logger.py', '2023-04-12 15:27:13.074792','[START] ptid: 110001','INFO', '44191')
CRITICAL DB ERROR! Logging to database not possible!: 1146 (42S02): Table 'log_db.detail_log' doesn't exist"

**Comment by rtdtwo** [comment]
@s-emerson I added the warning logs, let me check why the critical db error shows up. It never showed up to me but I am working on a database with tables already created, that could be a reason why.

**Comment by rtdtwo** [comment]
@s-emerson I did a run on a clean setup following the steps mentioned in this PR, however, I am unable to reproduce the issue.

**Comment by s-emerson** [comment]
When you get the time- how do I manually create the table? Are there any attributes I need to give it, or a specific name or anything?

**Comment by mbentz-uf** [comment]
This can be picked up again now that [report_handler](https://github.com/ctsit/report_handler) is ready

**Comment by s-emerson** [comment]
@sinarest1608 I ran into an issue when I tried to run nacculator after starting up the docker container and installing report_handler using "pip3 install ." . 

$ redcap2nacc -fvp <run_08-15-2023/final_Update.csv >run_08-15-2023/fvp_nacc_complete.txt
Traceback (most recent call last):
  File "/Users/s.emerson/Documents/Code/nacculator/.venv/bin/redcap2nacc", line 5, in <module>
    from nacc.redcap2nacc import main
  File "/Users/s.emerson/Documents/Code/nacculator/nacc/redcap2nacc.py", line 21, in <module>
    from nacc.uds3.ivp import builder as ivp_builder
  File "/Users/s.emerson/Documents/Code/nacculator/nacc/uds3/ivp/builder.py", line 9, in <module>
    from nacc.uds3 import clsform
  File "/Users/s.emerson/Documents/Code/nacculator/nacc/uds3/clsform.py", line 3, in <module>
    from nacc.logger import report_handler
ImportError: cannot import name 'report_handler' from 'nacc.logger' (/Users/s.emerson/Documents/Code/nacculator/nacc/logger.py)

It looks like this issue is caused by circular dependency, but I'm not sure where the issue would be caused.

File: nacc/logger.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import logging
+import sys
+from report_handler.report_handler import ReportHandler
+
+global report_handler
```
Comment (mbentz-uf) [inline nacc/logger.py:5]: The report handler is not being used as intended, it should not be a global variable. Define it outside of `configure_logging` and pass it in as an argument instead. Update `configure_logging` to take in an optional `List` of `logging.Handler`s i.e., `configure_logging(config, handlers: List[logging.Handlers])` and add them to the root logger. Then each entry point script should look something like:
```python
from nacc.logging import configure_logging
from report_handler.report_handler import ReportHandler

def entrypoint():
    report_handler = ReportHandler()
    configure_logging(config, [report_handler])
   
   # run nacculator logic
   # ...

   # after nacc finishes
   report_handler.write_report()
```

File: nacc/logger.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    # Gets the root logger and any config changes here affects logging across the code base
+    logger = logging.getLogger()
+
+    logger.addHandler(console)
+    logger.addHandler(report_handler)
+
+
+if __name__ == "logger":
+    configure_logging()
```
Comment (mbentz-uf) [inline nacc/logger.py:38]: Remove these lines.

**Review by mbentz-uf — CHANGES_REQUESTED** [review]
@s-emerson What are the entry point scripts for nacculator? For example in the case of pubingest, it's https://github.com/ctsit/ufvivo-pubingest/blob/master/ufvivo/pubingest/__main__.py.

@sinarest1608 Please clean up commented out code and make the necessary changes to `configure_logging`. Please let me know if you have any questions.

**Comment by s-emerson** [comment]
> @s-emerson What are the entry point scripts for nacculator? For example in the case of pubingest, it's https://github.com/ctsit/ufvivo-pubingest/blob/master/ufvivo/pubingest/__main__.py.


As far as I'm aware, the entry point script for nacculator is https://github.com/ctsit/nacculator/blob/master/nacc/redcap2nacc.py (which has a function "main" at the bottom). There's also the filters, which run using https://github.com/ctsit/nacculator/blob/master/nacc/run_filters.py (this is the script that does the api call to redcap and automatically downloads the project data).

File: nacc/logger.py (hunk 4 above)
Comment (sinarest1608) [reply nacc/logger.py:5]: @mbentz-uf 
I have added
```python
    report_handler = ReportHandler()
    configure_logging(config, [report_handler])
```
to  the main function at
> https://github.com/ctsit/nacculator/blob/master/nacc/redcap2nacc.py (which has a function "main" at the bottom).

**Comment by sinarest1608** [comment]
@mbentz-uf 
I have added the changes. 

@s-emerson 
Can you try running it again? You need not run the docker container since `Report Handler` does not require a mysql connection.

**Comment by s-emerson** [comment]
@sinarest1608 i tried running it again with the doctor container down and i got the same error.

**Comment by sinarest1608** [comment]
@s-emerson 
Can you list the steps to run nacculator so that I can try replicating the issue?

**Comment by s-emerson** [comment]
@sinarest1608 

with an API token from the REDCap project pasted into the nacculator_cfg.ini file and the redcap_server host set to our REDCap instance:

```
$ git checkout feature/ADRC-2460-add-db-logger
$ git pull
$ source .venv/bin/activate
$ pip3 install .
$ nacculator_filters nacculator_cfg.ini
$ redcap2nacc -fvp <run_08-24-2023/final_Update.csv >run_08-24-2023/fvp_nacc_complete.txt
```

though this also fails when you try to run `nacculator_filters nacculator_cfg.ini` with the issue:
AttributeError: module 'logging' has no attribute 'Handlers'

**Comment by sinarest1608** [comment]
@s-emerson 
1. From what project should I generate the API token to test this? Is there a test project for that?
2. How do I set this?

> the redcap_server host set to our REDCap instance

3. Would running this impact any data I should take care of?

**Comment by s-emerson** [comment]
i gave nacculator a try by using nacculator_filters to download from the redcap api and then running redcap2nacc with the -ivp, -fvp, and -m flags.

everything works fine, i didn't get any issues with the ivp_complete.txt or ivp_errors.txt files (and so on), which is great. so nacculator retains its previous functionality. 

however, when logging errors to the spreadsheet report-2023-09-08 (for example), i notice that several errors come back listed as "Unknown" instead of the error message that nacculator outputs to stderr. it looks like this is because they are caused by invalid data (like missing visit date) that don't normally get added to the list of errors that nacculator outputs and are instead printed by python itself. is there a way to also add these to the log?

i ran into a similar issue with the -m flag, except all of the errors were "Unknown" even when nacculator had a log message for the actual issue (milestone date was out of range, which i expected to see).

**Comment by sinarest1608** [comment]
@s-emerson 
I ran naacculator and it was terminated with this error
image

**Comment by s-emerson** [comment]
> @s-emerson I ran naacculator and it was terminated with this error image

This error is caused by an issue with the data and should be resolved. You might need to merge in the changes from develop in order for this to skip properly if it pops up again.

**Comment by sinarest1608** [comment]
> i gave nacculator a try by using nacculator_filters to download from the redcap api and then running redcap2nacc with the -ivp, -fvp, and -m flags.
> 
> everything works fine, i didn't get any issues with the ivp_complete.txt or ivp_errors.txt files (and so on), which is great. so nacculator retains its previous functionality.
> 
> however, when logging errors to the spreadsheet report-2023-09-08 (for example), i notice that several errors come back listed as "Unknown" instead of the error message that nacculator outputs to stderr. it looks like this is because they are caused by invalid data (like missing visit date) that don't normally get added to the list of errors that nacculator outputs and are instead printed by python itself. is there a way to also add these to the log?
> 
> i ran into a similar issue with the -m flag, except all of the errors were "Unknown" even when nacculator had a log message for the actual issue (milestone date was out of range, which i expected to see).

@s-emerson 
I looked into `report_handler` implementation and everything seems to be working as expected. I tried using different flags you mentioned and could see "Unknown" in place of the error messages. I looked through the implementation for `python_db_logger` and it also had the same placeholder for the error messages. I also checked the console logs and there is no specific error messages for these PTIDs. For example:
Screenshot 2023-10-30 at 7 15 15 PM

**Comment by sinarest1608** [comment]
@s-emerson 
Upon investigating further, I found that the errors with reason "Unknown" were actually exceptions (example the invalid data that you mentioned) that were begin caught but not printed. I have added the changes so that the report now has the exception messages. It should work as expected.

**Comment by s-emerson** [comment]
the code works well and the logs print out nicely in .xlsx format for all flags. 

one thing i notice is that the error printouts in the .xlsx log file don't have as much detail as the printouts in stderr. for example, for one ptid, the stderr printout said "Length of field DRUGID in form A4D with value "example error" is not valid. 11 != 6 which gives detail about the form the error is in, which field has the issue, and the value that is causing the issue. but in the log output, all that it says is "Assertion failed". 

can you have it print out the actual error message that nacculator gives? or is the .xlsx log supposed to be more general?

**Comment by s-emerson** [comment]
The code looks good and works well. Thank you, Kshitij!

**Pull request merged**